lib/BC/Util.cpp

```diff
@@ -437,4 +437,61 @@ void ConvertArrayArguments(llvm::Module &m) {
   CHECK(VerifyModule(&m)) << "Transformation broke module correctness";
 }
 
+static void FindRedundantLoads(llvm::Function &func) {
+  auto HasStoreBeforeUse = [&](llvm::Value *ptr, llvm::User *user,
+                               llvm::LoadInst *load) {
+    std::unordered_set<llvm::BasicBlock *> visited_blocks;
+    std::vector<llvm::Instruction *> work_list;
+    auto inst{llvm::cast<llvm::Instruction>(user)};
+    work_list.push_back(inst);
+    while (!work_list.empty()) {
+      inst = work_list.back();
+      work_list.pop_back();
+      auto bb{inst->getParent()};
+      visited_blocks.insert(bb);
+
+      while (inst) {
```

Comment:
Hrmm I can't see an easy way of doing that, because I also need to navigate through the predecessor blocks

Reply:
What I'm a bit worried about is that testing `while(inst)` that's moved by `inst = inst->getPrevNode()` is gonna run out of bounds. Hmm, what about instead of putting just the terminator into the worklist, you'd put in the whole predecessor BB in reverse via `work_list.push_back(pred.rbegin(), pred.rend())` ? And initialize via `inst->getReverseIterator()` and `inst->getParent()->rend()`.

Reply:
I know, but I'd still prefer using iterators, ranges and all that. I'm not gonna insist though.